### DISCUSSION TITLE
Error instantiating SQLite connection in WASM

### DISCUSSION THREAD

**Discussion by Suriman** [unoplatform/uno#12926] [Q&A]
### Current behavior

An error is produced when instantiating a SQLiteConnection in WASM head because it does not find e_sqlite3 dll.

### Expected behavior

SQLiteConnection is instantiated without errors.

### How to reproduce it (as minimally and precisely as possible)

Unzip the attached sample.
Open UnoApp1.sln
Run WinUI head, WPF head and/or Android head and the startup will complete and show the main page with a list of 3 items.
Run WASM head the startup will not complete and gets stuck in the splash screen. If you debug with breakpoints and activate all CLR exceptions, a DllNotFoundException is produced when executing SQLitePCL.Batteries.Init().
Open BlazorApp1.sln
Run and the startup will complete and show the main page with a list of 3 items.
[SqliteWasmIssue.zip](https://github.com/unoplatform/uno/files/12084092/SqliteWasmIssue.zip)

I think the problem is related to sqlite3.a file not being copied to the dist package (and I imagine also it has to be listed in some configuration files). In Blazor it is.

**Comment by jeromelaban** [comment]
Thanks for the report. You can find an example of application using SQLite [here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/EFCoreSQLiteSample/EFCoreSQLiteSample).

**Reply by ADD-David-Antolin** [reply]
Hello. I did review the sample and I have adapted my attached sample.

To make it work I have added these two NuGets to the WASM csproj:
```
<PackageReference Include="SQLitePCLRaw.provider.sqlite3" Version="2.1.2" />
<PackageReference Include="Uno.sqlite-wasm" Version="3.40.0.1" />
```
And also have added this if to the platform AppHead.xaml.cs before `base.OnLaunched(args)`:
```
#if __WASM__
    SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_sqlite3());
#endif
```

I'm glad it works now but my question remains the same. Blazor WASM gets it done without having to add those NuGets and that initialization code. I think you could achieve the same.

**Reply by jeromelaban** [reply]
@ADD-David-Antolin we welcome contributions on how to change that behavior, since SQLitePCLRaw is not part of Uno.

**Reply by ADD-David-Antolin** [reply]
@jeromelaban I'm glad to try, altough i will be doing it in my free time.

Here is what I have discovered by now:

First part, not needing to add
```
<PackageReference Include="SQLitePCLRaw.provider.sqlite3" Version="2.1.2" />
<PackageReference Include="Uno.sqlite-wasm" Version="3.40.0.1" />
```
Package [Microsoft.EntityFrameworkCore.Sqlite](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore.Sqlite/7.0.9) depends on [SQLitePCLRaw.bundle_e_sqlite3](https://www.nuget.org/packages/SQLitePCLRaw.bundle_e_sqlite3/) 
which depends on both [SQLitePCLRaw.lib.e_sqlite3](https://nuget.info/packages/SQLitePCLRaw.lib.e_sqlite3/2.1.4) and
[SQLitePCLRaw.provider.e_sqlite3](https://www.nuget.org/packages/SQLitePCLRaw.provider.e_sqlite3/2.1.4)

`SQLitePCLRaw.lib.e_sqlite3` contains `e_sqlite3.a` in path `runtimes/browser-wasm/nativeassets/net7.0`. 
That is the file that should get "bundled" in the dist package.  
Here in [Uno.Wasm.Bootstrap ShellTask](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/31fda3026df64623e1e5d85fa5d01600e9443f75/src/Uno.Wasm.Bootstrap/ShellTask.cs#L1669) it seems to do something similar to what it should be needed but it does only with the dotnet runtime.

It is my understanding that for Uno WASM if you set an item as `Content` it seems to get processed by `Emscripten`, so i have modified locally the file `SQLitePCLRaw.lib.e_sqlite3.targets` of package `SQLitePCLRaw.lib.e_sqlite3` in path `buildTransitive/net7.0` and added:
```
<ItemGroup Condition=" '$(UnoRuntimeIdentifier)' == 'WebAssembly' ">
    <Content Include="$(MSBuildThisFileDirectory)..\..\runtimes\browser-wasm\nativeassets\net7.0\e_sqlite3.a" />
</ItemGroup>
```
With that the sample runs (with some flaky OoM at loading) altough it seems to trigger AOT for all assemblies. 
Here are the binlogs: [content_workaround_binlogs.zip](https://github.com/unoplatform/uno/files/12113354/content_workaround_binlogs.zip)
 I imagine going full AOT for that is some kind of misbehavior. I do not know if those files should be managed in `ShellTask` with an `Item` different to `Content`. For the record blazor does it with `NativeFileReference`.

Second part, having to call manually:
```
SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_sqlite3());
```

That call is done inside `SQLitePCL.Batteries_V2.Init()`, [which is reflected-call inside the static constructor of SqliteConnection](https://github.com/dotnet/efcore/blob/5ae7876a3c3ca65b92ffc0c5387b74c9cacf2fb4/src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs#L51).
However, if a not explicit reference is made in code, `SQLitePCLRaw_batteries2.dll` gets trimmed and it is not present in the dist package. I have tried to add `<assembly fullname="SQLitePCLRaw.batteries_v2" />` to `LinkerConfig.xml` and also `<WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>` is in effect in (I am compiling in `Debug`) to no avail. Here is the binlog:
[sqlitepclrawbatteries_missing.zip](https://github.com/unoplatform/uno/files/12113457/sqlitepclrawbatteries_missing.zip)

**Reply by limeniye** [reply]
>Update Uno.sqlite-wasm3.40.0.2, it's great! Thanks

From reply: https://github.com/unoplatform/uno/discussions/14844#discussioncomment-8129866
